### ISSUE TITLE
Make it work on Mac/Linux	

### ISSUE THREAD

**Issue by activescott** [activescott/lessmsi#3]
from https://code.google.com/p/lessmsi/issues/detail?id=5

**Comment by john550** [comment]
is is possible to use lessmsi on linux.

**Comment by activescott** [comment]
Not yet, but the cab extraction code does all work on Linux now. So we're half way there. The part that need some work is the MSI code which is a combination of OLE Compound Files (aka Structured Storage) and a special MSI-specific database. I think the most reasonable thing to do is to use the WINE libraries (or their source?) to get the last part done. 

Always glad to help you along if you're interested in contributing!

**Comment by jgstew** [comment]
I have a very specific use case in which I need to get the MSI ProductVersion from the Property table of an MSI file using the command line. 

I still have not figured out how to do this using the command line and lessmsi on OS X, but I do have the lessmsi GUI working with WineBottler.  http://winebottler.kronenberg.org/

You can download the zip compressed .app I made here:  https://dl.dropboxusercontent.com/u/640753/lessmsi.zip

You can make it yourself using the following screenshot & winebottler:
![screen shot 2015-03-13 at 3 02 08 pm](https://cloud.githubusercontent.com/assets/2439367/6645360/4f9a462e-c992-11e4-8e02-a161096e3699.png)

**Comment by wyatt8740** [comment]
Just want to throw it out there that I've been able to use Less MSIerables in linux with wine after doing `winetricks dotnet40`. It's probable that wine's version of mono works as well - I can get a GUI to come up in the linux-native version but can't open MSI files due to the windows-specific code.

I think it's quite probable that it could be compiled with libwine, but don't have the hard disk space on this laptop to verify.

**Comment by jgstew** [comment]
I am able to use it on a Mac, and it doesn't have any windows-specific code outside of whatever is in Wine.

**Comment by wyatt8740** [comment]
well as wine consists of alternative API's for windows-only code, you're saying the wine team has done it's job well - not that it's actually a portable program. Right?

**Comment by jgstew** [comment]
Yes, Wine's code is windows specific, just not written by Microsoft.

LessMSI is not a portable program, but it works through Wine on the Mac, which doesn't have anything to help it run other than Wine and anything installed with Winetricks.

**Comment by pombredanne** [comment]
On my side I would be interested to have command-line only native (no wine required) way to build & run on Linux and Mac in addition to Windows, to add support for a better MSI extraction in https://github.com/nexB/scancode-toolkit beside the 7z-based extraction I have today.

**Comment by jgstew** [comment]
I haven't explored this option, but since lessmsi appears to be written in C# then it might be able to run on Mono or .NET core which would make it portable.

**Comment by qwertychouskie** [comment]
Has anyone tried compiling with `winelib`?  That should fill in all of the Win-specific functions.

**Comment by sskras** [comment]
@activescott wrote:

> The part that need some work is the MSI code which is a combination of OLE Compound Files (aka Structured Storage) 

Gnome `msitools` uses [GNOME/libgsf](https://github.com/GNOME/libgsf) which is partially influenced by Wine.

> and a special MSI-specific database. I think the most reasonable thing to do is to use the WINE libraries (or their source?) to get the last part done.

[GNOME/msitools](https://github.com/GNOME/msitools) used the same approach, IIUC. 

Both libs are under LGPL-2+. For MIT-licensed project it would be hard to reuse GPL code directly.

**Comment by jgstew** [comment]
I've actually switched to using https://github.com/GNOME/msitools for my non-windows use cases for this. That said, I wish I had the time to experiment with lessmsi to see how easy it would be to get it to work with winelib or mono.

**Comment by def-fun** [comment]
> I've actually switched to using https://github.com/GNOME/msitools for my non-windows use cases for this. That said, I wish I had the time to experiment with lessmsi to see how easy it would be to get it to work with winelib and/or mono.

@jgstew thanks for your share. can you show me how to deploy msitools on linux?

**Comment by nightlark** [comment]
Thanks for catching the typo!

It’s using Pyodide, a version of CPython compiled to Webassembly. Generally it works with pure Python libraries (no native code/dependencies), which was one of the requirements we had for pymsi. Beyond that, the UI is HTML/JavaScript, with some calls to run some small snippets of Python code that act as a bridge for calling pymsi functions to handle working with the MSI file contents.
